Pantheon Provider plugin

beeradb opened the pull request:
## What happened (or feature request):
As per #216 we need pantheon integration. This PR introduces the integration.

At this point, it is feature complete enough to take it for a test drive.

There are three new commands which will be needed:
- `ddev config-pantheon [token]` sets global configuration for talking to the pantheon API. Please see https://pantheon.io/docs/machine-tokens/ for documentation on creating a machine token.
- `ddev config pantheon` extends standard ddev config to account for pantheon specific data.
- `ddev import` imports from a provider. It prints an error if no external provider has been configured.


There are also two new files which get added to the `.ddev` folder:

- import.yaml - This contains pantheon provider-specific data. The goal is that all providers would use this same import.YAML (although with different fields).
- provider.env - This is an environment file included by default in docker-compose. It includes provider specific environment variables. In the case of pantheon, Pressflow allows for various configuration options using environment variables, and those get written to provider.env.

## How to reproduce it (as minimally and precisely as possible):

1. Deploy a new site to pantheon. Either Drupal or WordPress will do.
2. After installation, switch the site to "git" mode and create a backup of the site.
3. Git clone the repository.
4. Run `ddev config pantheon` from the git root directory. Follow configuration prompts.
5. Run `ddev import`

## Related source links or issues:
#216 is the parent issue.

beeradb commented:
This has been rebased and is "test drive" worth at this point.

beeradb commented:
## Additional architectural notes

There are a couple decisions here which I'm only somewhat sure I like. I fiigured I may as well call them out.

- Provider is an interface in the `ddevapp` package. I'm not sure how useful having an interface that only has with only a single (valid) implementation. It seems inevitable to me that if we add more than a single provider we'll need the interface, but it also may change drastically as we add more providers.

- There is a empty "DefaultProvider" struct which implements the provider interface but no-ops all the methods. This seemed preferable to me than always gaurding calls to provider methods with if statements to determine if a valid provider config was found or not, but it does feel a bit strange.

rickmanelius commented:
Testing https://github.com/drud/ddev/pull/345

## Reference

* [MVP Integration w/Pantheon](https://github.com/drud/ddev/issues/216)
* [Pull Request 345](https://github.com/drud/ddev/pull/345)
* [Circle CI ddev 1201](https://circleci.com/gh/drud/ddev/1201#artifacts/containers/0)

## Preparation

* Download ddev_osx.v0.7-33-ga5397661.tar.gz
* Place in /usr/local/bin/ddev
* `ddev version` outputs v0.7-33-ga5397661

## Testing

* Downloaded a pantheon repo.
* Attempted to run `ddev config pantheon` and was shown an error:
  * "2017/06/29 11:59:31 No saved session could be found and the environment variable TERMINUS_API_TOKEN is not set. Please set it to a valid Terminus API Token."
  * Might want to actually call out how to create / obtain the token.
* Attempted `ddev config-pantheon` and received the following error:
  * "You must provide a Pantheon  machine token, e.g. `ddev config-pantheon [token]`. See https://pantheon.io/docs/machine-tokens/ for instructions on creating a token."
  * Might just want to prompt them.
* Ran `ddev config-pantheon MYTOKEN`
  * Success!
* Re-ran `ddev config pantheon`
  * Configure import environment shows 3 options, but I'm not clear if I just need to type it out.
  * Might want to default to `dev` as a suggestion.
  * Putting in an incorrect value (test2) did result in the command prompting me again.
  * I see "You may now run `ddev start` or `ddev import`"
* Run `ddev import`
  * I see the new containers running.
  * I see the import of db (6 MB) and files (130 MB)
  * I see the "import success, restarting"
  * I don't see a URL appear. That would be handy.
  * The entire load time from end to end was 1:45 (this included the containers)
* Site loads!!!
* Re-run `ddev import`
  * I receive no warning that I'm about to nuke my db! :)
  * I see no info about a database or files download occuring.
    * After re-reading the pull request thread, I see this is a baked in feature. Nice!
  * Reimport worked!
    * I'm assuming it's using a cached copy?
* I re-ran a backup job on my live site to create a newer backup.
  * The download time was about 50 seconds of the entire 1 minute process.

## Spitballing

* What if someone wanted to override the default source? Maybe `ddev import dev`. This is by no means necessary now, but might be a future enhancement.
* What if someone just wanted to pull the database and not files?

rickmanelius commented:
Overall, my reaction is very positive. There is a bit of polish and/or a bit of guidance that can be provided to assist. Example, running the import without a `ddev start` does spin up the containers, but the URL is not exposed/printed at the end. Little enhancements like that would go a long way.

There is also a bigger need (which I can understand why we would wait for feedback at the technical level first) of getting information ready for our documentation. That in no way impacts the technical review and some of that is included in this PR summary. However, we'd want to make sure this doesn't get neglected.

rickmanelius commented:
One other error that popped up. When running `ddev import` a few times in a row, I did hit this once.
`Could not perform import: Get https://terminus.pantheon.io:443/api/sites/ac84a9a6a3-0886-4876-a4b7-8b53b5c36dbc4d/environments/live/backups/catalog: EOF`

beeradb commented:
I made the following updates based on feedback from @rickmanelius 

- Linked to https://pantheon.io/docs/machine-tokens/ in all places where help text/errors are output do to a token not being provided.
- Changed the environment selection during config from "please choose an environment" to "Type the name to select an environment to import from"
- Added a `--skip-confirmation` flag + warning prompt to `ddev import` to alter users of potential data loss.
- Added URL output at the end of `ddev import`

I did NOT do the following:
- Prompt user for token value when not providing one via `ddev config-pantheon [token]"

I think prompting works best to avoid flag/argument soup, such as in `ddev config` where we need to capture a range of values. I'm not very sold on it being a good idea when we have a single value being configured. I'm not completely against it, but consider this some gentle pushing back against the idea :)

rickmanelius commented:
I'm mid review on the updates. One thing I think that has been surfaced before and I feel strongly will bite us in the ass is that "ddev config-pantheon" and "ddev config pantheon" differ by only a hyphen and yet do completely different things. We discussed things like "auth" or "connection" for the global config. See notes here https://github.com/drud/ddev/issues/216#issuecomment-307220118.

rickmanelius commented:
When running `ddev import`, I had the following appear, which I believe is not what is anticipated.
`%!(EXTRA string=pantheonsitename)`

beeradb commented:
This is because I updated the warning string without removing the appname parameter. Oops.

## Brought to you by

![orly](https://pics.me.me/just-put-the-technical-debt-on-my-credit-card-moving-18535009.png)

rickmanelius commented:
So yeah. The latest changes are very helpful. Still a nit about the global config name and the string error cropped up. Beyond that, I think from the end-user perspective this is worthy of a merge, but it does need a technical review and docs. I'm cool with splitting out documentation as a second PR if it would be a big lift (trying to keep commits smaller so that we don't keep long standing WIP).

beeradb commented:
Agreed. I do hate `ddev config-pantheon` vs `ddev config pantheon`. It's something that bothered me when developing, but I figured I'd wait and see what kind of feedback I got.

I'd vote to move it to `ddev auth-pantheon` since (at this point) that's all it does. We have no guarantee we won't ned more config for a future provider, but we can always revisit when we have that information.

rickmanelius commented:
I'm good with `ddev auth-pantheon` knowing that it may change and is an accurate placeholder.

rickmanelius commented:
As discussed in Slack, I'll take a first stab at the documentation.

rickmanelius commented:
Documentation is being addressed here https://github.com/drud/ddev/issues/369 and is in a "Needs Review" status.

rickmanelius commented:
@rfay @tannerjfco When you get a chance, I would love to get reviews on this item to get to a consolidated punch list. Thanks!

tannerjfco commented on pkg/ddevapp/providerPantheon.go:
```diff
@@ -0,0 +1,365 @@
+	}
+
+	if p.Environment == "" {
+		p.Environment = "dev"
```
"live" seems like the most appropriate default, but i suppose with pantheon that doesn't exist until a site is close to done (and a credit card has been provided).

beeradb replied:
Yep! I chose this as the default for that very reason. Dev is the first environment created by default on pantheon, and as such feels like the most consistently available.

rfay replied:
I think dev is the right default in every case

tannerjfco commented on pkg/ddevapp/providerPantheon.go:
```diff
@@ -0,0 +1,365 @@
+		keys = append(keys, k)
+	}
+	fmt.Println("\n\t- " + strings.Join(keys, "\n\t- ") + "\n")
+	var environmentPrompt = "Type the name to select an environment to import from"
```
What about numbered options for this prompt?

tannerjfco commented on pkg/ddevapp/providerPantheon.go:
```diff
@@ -0,0 +1,365 @@
+		// If we can't read a previous session fall back to using the API token.
+		apiToken := os.Getenv("PANTHEON_API_TOKEN")
+		if apiToken == "" {
+			log.Fatal("No saved session could be found and the environment variable PANTHEON_API_TOKEN is not set. Please use ddev config-pantheon or set a PANTHEON_API_TOKEN. https://pantheon.io/docs/machine-tokens/ provides instructions on creating a token.")
```
util.Failed

tannerjfco commented on pkg/plugins/platform/local.go:
```diff
@@ -423,14 +473,12 @@ func (l *LocalApp) Start() error {
-		if err != nil {
-			return err
-		}
+	err := l.AppConfig.WriteDockerComposeConfig()
```
are we going to 'own' the compose file with this? We may want to consider using the file signature solution randy implemented here if so.

beeradb replied:
Nope! the write is still conditional, the logic was just moved inside the function.

tannerjfco commented on pkg/util/network.go:
```diff
@@ -30,9 +32,20 @@ func DownloadFile(filepath string, url string) (err error) {
 		return fmt.Errorf("download link %s returned wrong status code: got %v want %v", url, resp.StatusCode, http.StatusOK)
 	}
+	reader := resp.Body
+	if progressBar {
```
shiny!

beeradb replied:
indeed! I felt like a progress bar was a must, as some archives might be quite large, and offering no visual feedback during a 5 gig download seems like a non-starter.

beeradb commented on pkg/ddevapp/config.go:
```diff
@@ -212,22 +284,40 @@ func (c *Config) Hostname() string {
-	if err != nil {
-		return err
+		// Create an empty provider.env file to make docker-compose happy.
+		_, err := os.Create(c.GetPath("provider.env"))
```
You will get an empty `provider.env` file regardless of whether you are using pantheon or not.  Otherwise, docker-compose will throw an error when using the `env_file` directive. An empty provider.env is now created alongside the docker-compose.yaml, so the user can commit it or ignore it as they choose.

I don't love this, but we need a way to conditionally include environment vars. Open to other suggestions on how to achieve it.

tannerjfco replied:
An alternative would be to write a docker-compose.pantheon.yaml that sets the environment variables (or adds an env_file directive). This could potentially be a file managed by ddev.

beeradb commented on cmd/ddev/cmd/import.go:
```diff
@@ -0,0 +1,65 @@
+
+		client := dockerutil.GetDockerClient()
+
+		err := dockerutil.EnsureNetwork(client, netName)
```
This is repeated all over the place, my next commit refactors into a single `dockerNetworkPreRun()` function for consistency. It also changes the error reporting to use `util.Failed()`

beeradb commented:
> Is it worth peeking in ~/.terminus/cache/tokens to get the token with config-pantheon? Or could we just use that cache instead of creating .ddev/pantheonconfig.json?

We could probably peek if we wanted. Terminus does support multiple logins in that folder, though, which could raise some concerns over which we use (just randomly pick one? try multiple to see if they are active? what if it's an old account or an expired token?). We'd need to decide what the behavior should be.

Personally, I'm fairly against writing anything to the terminus token cache.

> I really can't go with competing ddev config pantheon and ddev config-pantheon commands. 

Agreed. `config-pantheon` has been renamed to `auth-pantheon` in the latest push.

> Why does it go through the whole start (container creation) process just to do it again after the import?

Containers need to be started to perform the additional import. After the import is done, we need to inject DB credentials and other settings via environment vars (since settings files are committed on pantheon). Doing this requires us writing a provider.env (or perhaps a  docker-compose.pantheon.yaml as Tanner suggested earlier) and a restart of the site for those variables to take effect.

> ddev import seems like a problematic addition. I think it will be confusing.

Totally open to alternative suggestions. We could re-use import-db and import-files and break it into a two-step process, but I was worried about overloading those commands with vastly different behaviors.

tannerjfco commented:
> > Why does it go through the whole start (container creation) process just to do it again after the import?

> Containers need to be started to perform the additional import. After the import is done, we need to inject DB credentials and other settings via environment vars (since settings files are committed on pantheon). Doing this requires us writing a provider.env (or perhaps a  docker-compose.pantheon.yaml as Tanner suggested earlier) and a restart of the site for those variables to take effect.

Do we need to read values from the imported database (or some other downloaded asset) in order to set these env vars? Just wondering if we can set these vars before the import occurs, so that we don't have to restart.

> > ddev import seems like a problematic addition. I think it will be confusing.

> Totally open to alternative suggestions. We could re-use import-db and import-files and break it into a two-step process, but I was worried about overloading those commands with vastly different behaviors.

It makes sense to me to re-use the separated import commands. Dealing with locally stored assets, their functionality isn't all that impressive, but I think they become quite a bit more impressive and useful handling syncing from upstream environments. I also think developers should be able to import just the database or files, regardless of provider. It's not uncommon for developers to only want a db dump, and to use something like stage_file_proxy to avoid pulling files (this becomes necessity for sites w/ a large volume of assets). I don't feel like it would be unreasonable for these to accept provider as argument, e.g. `ddev import-db pantheon`. I'm not opposed to a command that does both db and files, but I think any import commands should work both locally or with a provider.

tannerjfco commented:
We do still need to set env vars for these WP salts. Without them set, wp-cli throws errors and outputs the wp-config.php contents with each command executed.

AUTH_KEY
SECURE_AUTH_KEY
LOGGED_IN_KEY
NONCE_KEY
AUTH_SALT
SECURE_AUTH_SALT
LOGGED_IN_SALT
NONCE_SALT

Not sure what we're doing for Drupal's hash salt, but I think we're safe to random-generate these.

rfay commented:
Request to @rickmanelius for tomorrow: Propose changing `ddev import` for `ddev pull` because of the collision and confusion with "import" vs "import-db"/"import-files".

rickmanelius commented:
# Proposed Changes

## Quick Summary

* Rename `ddev import` with `ddev pull`.
* File an issue with `ddev push` as a placeholder for future functionality.
* File an issue to add the following flags to `ddev pull` for future functionality.
  * --ignore-db (does not download a new database archive and uses most recent version)
  * --ignore-files (does not download a new file archive and uses most recent version)
  * --skip-import (performs the download of files and/or db but does not perform the import)
  * --env=ENV (allows users to target a different environment than the one specified in config)
* Ship the initial pantheon plugin as is with this single change in place (e.g. s/import/pull/)


## Discussion

At a high level, what we are basically discussing is downloading/uploading archives (db, files, both, neither), to or from a specific environment, and then importing them (db, files, both, or neither). Sounds simply, but that represents at least 4 * 4 combinations of options of "pull" IF pull/push are assumed to take on both the moving and importing of the archive by default.

Right now, we collapse this significantly because `ddev import pantheon` assumes one downloads and imports both. But as @tannerjfco has noted, some users probably don't want to re-download the file assets over and over again. And given that Pantheon provides the archives separately, you could do that.

If we limit the scope of "pull" to just pulling assets and then require a follow-up import, that would simplify things greatly but be kindof annoying/tedious because I believe the more typical use case would be to do both and requiring users on the CLI to stack each time would be tedious.

Getting around this, `ddev pull pantheon` would by default read the source environment and then attempt to download new assets from that environment (unless they were already up to date) and then import. Alternatively they could exclude files by adding --ignore-files, which would skip the download and then import from a local cache.

Alternatively, we can keep the assumption that any pulled asset would be then imported and "--ignore-db". So essentially if we're saying not to download then we're also saying not to import. The only time that wouldn't make sense is if you wanted to use a local cached copy and yet still import.

## Being Pragmatic

Ultimately, I think we start with the assumption of pulling and importing both and then extend later on. There are some use cases (specifically former partners like newmedia that use full drush archives and therefore they can't target downloading the db without the files), that this is moot anyway from a pattern perspective.

So that is why I suggest we rename for now. That way it gives users are clearer sense that pull/push means the transfer of files and is less likely to get conflated with the utility commands of import-db and import-files, which are performing the heavy lifting locally regardless of if the archive came from somewhere else or is being imported manually. And the pull command is still platform specific because it will know the expected archive locations and formats and pipe that information to import-db and import-files. So each set of commands is still valuable and not to be confused with each other.

beeradb commented:
I've added the additional Wordpress settings. Unfortunately, wp-cli still complains quite loudly, do to the standard pantheon `wp-config.php` defining WPLANG, which wp-cli also does.

I've opened this PR against the pantheon repo, but it's out of control for now.

https://github.com/pantheon-systems/WordPress/pull/127

beeradb commented:
I've updated this PR to remove the prepend.php technique from https://github.com/drud/docker.nginx-php-fpm-local/pull/29 in favor of using settings.local.php (for Drupal) and wp-config-local.php (for WordPress).

I believe, at this point, that all feedback provided has been accounted for. I'm moving back to working towards having the test coverage needed to merge this.

This PR is ready for another round of feedback, or we can just wait until test coverage is where we want it (I'm guessing tomorrow afternoon).

rickmanelius commented:
@beeradb Thanks! I'm happy to review now versus waiting until all test coverage is there in case of anything changing as a result of the feedback (not likely, but I'd rather get ahead of it).

tannerjfco commented on cmd/ddev/cmd/pull.go:
```diff
@@ -0,0 +1,60 @@
+var PullCmd = &cobra.Command{
+	Use:   "pull",
+	Short: "Import files and database using a configured provider plugin.",
+	Long: `Import files and databased using a configured provider plugin.
```
s/databased/database

rickmanelius commented:
Previous Review https://github.com/drud/ddev/pull/345#issuecomment-312053813

## Preparation

* git checkout 1b39d48d
* make clean && make darwin
* ddev version outputs v0.8-40-g1b39d48d
* Removed ~/.ddev/pantheon
* Removed ~/.ddev/pantheonconfig.json
* Removed ~/git/pantheon-site (after running `ddev remove pantheon-site`)

## Testing

* Downloaded a pantheon repo.
* Attempted to run ddev config pantheon and was shown an error:
  * No saved session could be found and the environment variable PANTHEON_API_TOKEN is not set. Please use ddev auth-pantheon or set a PANTHEON_API_TOKEN. https://pantheon.io/docs/machine-tokens/ provides instructions on creating a token.
* Attempted `ddev auth-pantheon` and received the following error:
  * You must provide a Pantheon machine token, e.g. `ddev auth-pantheon [token]`. See https://pantheon.io/docs/machine-tokens/ for instructions on creating a token.
* Attempted `ddev auth-pantheon INVALIDTOKEN` and received the following ERROR:
  * Could not authenticate with pantheon: 500 Internal Server Error: 500
* Attempted `ddev auth-pantheon VALIDTOKEN` and received the following SUCCESS:
  * Authentication successful! You may now use the `ddev config pantheon` command when configuring sites!
* Re-ran ddev config pantheon
  * Walked through normal configuration options.
  * Regarding import environment, it defaulted to `dev`
  * Attempted to put an invalid environment (live 2) and it wouldn't let me through.
  * Saw follow-up instructions on `ddev start` and `ddev pull`
* Ran `ddev pull`
  * Saw a warning about a database that doesn't exist! :)
  * The import failed with the following:
    * "Could not perform import: could not find a backup of type database. please visit your pantheon dashboard and ensure the 'dev' environment has a backup available"  
  * Created a backup in the dashboard.
  * Re-ran `ddev pull`
  * I see the import of db (6 MB) and files (130 MB)
  * I see the "import success, restarting"
  * I attempt to view the site and see the following error:
    "ddev web container: 500 or 501: Internal Server Error or Unimplemented"
* I tried a `ddev stop` and `ddev start`, no dice.
* I tried a `ddev pull` and reimport, no dice.

FWIW, I also had 3 other sites still active, so I'm not sure if there is something at the nginx proxy level that needs to be reset.

tannerjfco commented on pkg/ddevapp/config.go:
```diff
@@ -408,17 +496,20 @@ func determineAppType(basePath string) (string, error) {
 	switch appType {
 	case "drupal8":
 		fallthrough
 	case "drupal7":
```
(not a blocking comment) Perhaps we should do a map[string]string of settings files? Could maybe let us handle the drush settings file better than we are now too.

tannerjfco commented on pkg/plugins/platform/local.go:
```diff
@@ -291,6 +288,52 @@ func (l *LocalApp) SiteStatus() string {
+		return err
+	}
+
+	util.Success("\nImport successful, restarting application.\n")
```
It doesn't seem like we restart here anymore, message could use an update here.

tannerjfco commented on pkg/plugins/platform/local.go:
```diff
@@ -291,6 +288,52 @@ func (l *LocalApp) SiteStatus() string {
+	}
+
+	if l.SiteStatus() != SiteRunning {
+		util.Warning("Site is not currently running. Starting site before performing import.")
```
Might be just me, but this seemed weird as a warning. I would be ok with plain text personally.

beeradb replied:
I wasn't really thinking about it as "warning", more as "visible yellow text". Seemed worth informing the user we were starting the site.

That being said, I agree and am going to roll it back. We should probably doing things that would confuse our usage of warning/success/failed.

tannerjfco commented on pkg/plugins/platform/local.go:
```diff
@@ -291,6 +288,52 @@ func (l *LocalApp) SiteStatus() string {
 	return siteStatus
 }
 
+// Import performs an import from a local
```
Could use a complete comment

rickmanelius commented:
Picking this back up for another turn.

rickmanelius commented:
Alright. I can confirm that this process correctly produces a settings.local.php file that can get the site into a functional state if/when one includes the following at the bottom of a settings.php file.

```
/**
 * Include a local settings file if it exists.
 */
$local_settings = dirname(__FILE__) . '/settings.local.php';
if (file_exists($local_settings)) {
  include $local_settings;
}
```

The challenge is that, unlike the prepend.php solution that was tried earlier, I as the end-user didn't necessarily know to go in and manually connect up that local settings file. This adds a further complication of whether or not we append the settings.php or try to detect if there is already a code snippet to include a local settings file.

rickmanelius commented:
To summarize the source of the issue and how we plan to address.

* Surface level issue. While ddev was generating a settings.local.php file with the correct credentials, it was not being referenced in settings.php and being ignored.
* Deeper issue. While Drupal 8 has standardized the inclusion of the settings.local.php file at the end of the default settings.php file, this has not been backported to Drupal 7. Additionally, Pantheon doesn't necessarily care if one commits a settings.php file, which ddev then recognizes as a custom settings.php and thus leaves it alone.

We have additional open issues that touch on this. Notably a directly related request here #346 and an indirectly related pattern here #343.

For the purposes of moving this forward, we discussed identifying when a user had a committed settings.php file without an inclusion of settings.local.php and providing them instructions on how to add that to their committed settings.php. We can then follow-up on these other issues and potentially solve it in a more robust fashion.

tannerjfco commented:
Bug from testing: If I run `ddev config` (no provider arg) for a site, then run `ddev config pantheon`, I do not get prompted for a pantheon environment, and no import.yaml is written. I still get the "You may now run `ddev start` or `ddev pull`" message on completion. 

This seems to relate to `provider: default` being written to the config.yaml, if I remove that line and run `ddev config pantheon` again, i get the pantheon prompts and import.yaml as expected.

tannerjfco commented:
I wonder if we should prompt for pantheon site name separately. The values are stored separately, and I could see how I might want to give one name for a site locally and pull from a different pantheon name. It seems like you can do this fine w/ manual edits after configuration, so not a necessity, but it might be helpful to prompt pantheon site name and default it to the base site name.

beeradb commented:
I'm fairly against adding a second prompt for Pantheon site name, personally. I'd really like to keep this as streamlined as possible, and when doing a checkout from Pantheon, the site name correctly defaults based on the directory name of the checkout. As @tannerjfco pointed out, there is nothing preventing a user from manually configuring a site to have a different name, but it feels like an edge case. I'd prefer to keep the configuration as streamlined as possible.

tannerjfco commented:
I took this for a test run in Windows as well. You need the db container fix from drud/docker.mysql-local#25, but with that in place everything seemed to work as I expected. The only thing I noticed is that the progress bar doesn't seem to work out if the Command Prompt window isn't wide enough - output below is the window at its default size when opened. If I maximize the window, I get the progress bar. Probably not worth worrying about for now. Everything else I tested seemed to work as expected!

```
C:\Users\tannerjfco\Downloads\tannerpress5000>ddev pull
You're about to delete the current database and files and replace with a fresh import. Would you like to continue (y/N):
y
C:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_database.sql.gz 0 B / 125C:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_database.sql.gz 67.57 KiBC:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_database.sql.gz 125.65 KiB / 125.65 KiB  100.00% 0s
Importing database...
Generating wp-config-local.php file for database connection.
Wordpress sites require a search/replace of the database when the URL is changed. You can run "ddev exec 'wp search-replace [http://www.myproductionsite.example] http://tannerpress5000.ddev.local'" to update the URLs across your database. For more information, see http://wp-cli.org/commands/search-replace/
C:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_files.tar.gz 0 B / 1.57 MC:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_files.tar.gz 61.08 KiB / C:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_files.tar.gz 214.08 KiB /C:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_files.tar.gz 503.08 KiB /C:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_files.tar.gz 1.14 MiB / 1C:\Users\tannerjfco\.ddev\pantheon\tannerpress5000\tannerpress5000_dev_2017-07-11T20-45-52_UTC_files.tar.gz 1.57 MiB / 1.57 MiB  100.00% 0s
Importing files...

Import successful, restarting application.

Successfully Imported.
Your application can be reached at: http://tannerpress5000.ddev.local
```

beeradb commented:
Remaining hitlist:

- [ ] Improved documentation for test dependencies
- [ ] Fix code review feedback from @tannerjfco
- [ ] Consider moving import.yaml fields into config.yaml (??)
- [ ] Consider a full site import test for pantheon.
- [ ] Add a notification if we can't write to settings.php to notify the user to include settings.local.php (??)

rickmanelius commented:
@beeradb As discussed this morning, I do know we have unit tests on the download and the import portion. Not sure we need we need to integration test that component, but it would be a nice to have.

tannerjfco commented:
> Consider moving import.yaml fields into config.yaml (??)

I'm in favor of this, but wouldn't block on it, especially if its significant effort at this point.

> Consider a full site import test for pantheon.

I'm not too concerned about this personally, especially if we have good test coverage on the pantheon package.

> Add a notification if we can't write to settings.php to notify the user to include settings.local.php (??)

I think it would be reasonable to notify if we can't write the settings file, that the user needs to provide valid configuration details from ddev describe.

rickmanelius commented:
## Previous Reviews

* https://github.com/drud/ddev/pull/345#issuecomment-312053813
* https://github.com/drud/ddev/pull/345#issuecomment-314275490

## Preparation

* git fetch --all
* git checkout fa2f6c78
* make clean && make darwin
* ddev version outputs v0.8-21-gfa2f6c78 (matches above)
* `ddev rm pantheon-site`
* `rm ~/git/pantheon-site`
* `rm -rf ~/.ddev/pantheon-site`
* `rm -rf ~/.ddev/pantheon`
* `rm ~/.ddev/pantheonconfig.json`

## Testing

* Downloaded a pantheon repo via git clone method.
* Attempted to run `ddev config pantheon` and was shown an error:
  * No saved session could be found and the environment variable PANTHEON_API_TOKEN is not set. Please use ddev auth-pantheon or set a PANTHEON_API_TOKEN. https://pantheon.io/docs/machine-tokens/ provides instructions on creating a token.
* Attempted `ddev auth-pantheon` and received the following error:
  * You must provide a Pantheon machine token, e.g. `ddev auth-pantheon [token]`. See https://pantheon.io/docs/machine-tokens/ for instructions on creating a token.
* Attempted `ddev auth-pantheon INVALIDTOKEN` and received the following ERROR:
  * Could not authenticate with pantheon: 500 Internal Server Error: 500
* Attempted `ddev auth-pantheon VALIDTOKEN` and received the following SUCCESS:
  * Authentication successful! You may now use the `ddev config pantheon` command when configuring sites!
* Re-ran ddev config pantheon
  * Walked through normal configuration options.
  * Regarding import environment, it defaulted to `dev`
  * Attempted to put an invalid environment (live 2) and it wouldn't let me through.
  * Saw follow-up instructions on `ddev start` and `ddev pull`
* Ran `ddev pull`
  * Saw a warning about a database that doesn't exist! :)
  * I see the import of db (6 MB) and files (130 MB)
* I attempt to view the site and see the following error:
  * "ddev web container: 500 or 501: Internal Server Error or Unimplemented"
  * We had discussed notifying an end-user that if they already had a custom settings.php file that they would need to add the settings.local.php include at the end of their settings.php file. See https://github.com/drud/ddev/pull/345#issuecomment-314527971


## Recommendation

* Print instructions to the end-user on how to fix a custom settings.php file.
* Technically when running `ddev config pantheon`, the check for the API key happens after asking for the project name. I wonder if that's worth checking for first because why ask for any information if it's going to abort?
* Instruction clarification. When we suggest using `ddev auth-pantheon` in the instruction "Please use ddev auth-pantheon or set a PANTHEON_API_TOKEN", do we want to make it a point to specifically call out command line instructions by wrapping in ``? Example, in the subsequent error message when running `ddev auth-pantheon`, the follow-up instruction notes `ddev auth-pantheon [token]`. This is not only more accurate (because the `auth-pantheon` command requires the additional argument and it is more obvious that it's a command with the surrounding `` characters).
* When running the first pull, we see the standard warning "You're about to delete the current database and files and replace with a fresh import. Would you like to continue (y/N):". Not sure how easy it is to check if there is a database or files to overwrite and therefore skip the warning the first time or provide a different message "no files or database detected so import will begin automatically."

rickmanelius commented:
I'm also in favor of notifying users about the settings.php file changes. If I didn't already know what to do, I would be SOL.

beeradb commented:
> I'm also in favor of notifying users about the settings.php file changes. If I didn't already know what to do, I would be SOL.

I'm in favor of doing that but I would much rather do it in a followup than hold up this PR for it. We've already made significant improvements to the settings file writing in this PR, and although more are possible (and encouraged!) I really want to set boundaries on this PR, as the surface area is large enough that I think we could add "just one more thing" indefinitely.

To recap: 
The settings file changes introduced here are already an improvement over what is in previous releases, and we should continue to iterate & improve, just not here, please! :)

rickmanelius commented:
@beeradb I believe both @tannerjfco and @rfay have already done multiple code and manual testing reviews. As soon as one/both of them gives a green, let's get this feature merged!

rfay commented:
Thanks so much for all the work on this. I've approved, but have to confess that I haven't tested the updated version nor have I reviewed the code yet. I do hope to do at least one of those this weekend. IMO it's quite quite late for code review, so if I do that I'll be gentle.

rfay commented:
Let's do make sure that the known followup issues are created before pull, just as a matter of discipline. Maybe we have them all in place.

rfay commented on cmd/ddev/cmd/config.go:
```diff
@@ -35,6 +52,12 @@ var ConfigCommand = &cobra.Command{
+
+		// If a provider is specified, prompt about whether to do an import after config.
+		if provider != "" {
+			util.Success("Configuration complete. You may now run `ddev start` or `ddev pull`")
```
This should not have backticks. Sorry, a nit, but this output is not markdown. Please check all of these to remove the backticks, which are just distracting to a normal user. Quotes would be OK, like "you may now run "ddev start" or "ddev pull"

beeradb replied:
I'm going to make these changes but it feels like it may be something worth discussing at some point. I agree that this is not markdown and we don't get improved formatting from this, but in many places (markdown, slack chats, shell, gitter chats) backticks are used to indicate commands or shell output.

I can accept if I'm the only one, but to me at least, seeing backticks around text makes it much more clear that it's an intended command than quotes do.

beeradb replied:
Scratch that, I am actually not making this change in this PR.

I find it clear as is, but I'm still happy to change it if it's something you and/or others are passionate about, but a quick grep for "`ddev" shows 80 occurrences of this pattern, including many places not touched by this work. If we make this change, we should make it everywhere for consistency, and doing that across so many files is probably best to do in a follow-up with clear intent, rather than just adding noise to the work here.

rfay replied:
I'm ok with a followup. Backticks are definitely not the way to communicate with our users. And we absolutely do need punctuation, I think people would expect double or single quotes, which are what was always used pre-markdown (and by people who don't use markdown).

Note that I would assume lots of the things you found in the grep were in our docs where they *do* belong.

rfay commented on pkg/archive/archive.go:
```diff
@@ -134,6 +134,13 @@ func Untar(source string, dest string, extractionDir string) error {
+			fullPathDir := filepath.Dir(fullPath)
+			err = os.MkdirAll(fullPathDir, 0755)
+			if err != nil {
+				return err
```
This would be better with context added via wrapping or fmt.errorf()

rfay commented on pkg/ddevapp/config.go:
```diff
@@ -243,22 +315,28 @@ func (c *Config) Hostname() string {
+			return err
+		}
+	}
 	return err
```
If I'm not mistaken, this would be clearer by removing the declaration of err at the top and then returning nil here. Because err is not going to be an error right? (and err is shadowed in every stanza above)

rfay commented on pkg/ddevapp/pantheon_test.go:
```diff
@@ -0,0 +1,130 @@
+
+	// testcommon.Chdir()() and CleanupDir() checks their own errors (and exit)
+	defer testcommon.Chdir(testDir)()
+	defer testcommon.CleanupDir(testDir)
```
The order of the defers has to be swapped here, will fail on windows. Since defers are LIFO, this tries to remove the dir before chdir-ing out of it. Of course by now /me completely hates this pattern :)

rfay commented on pkg/ddevapp/pantheon_test.go:
```diff
@@ -0,0 +1,130 @@
+	assert.NoError(err)
+}
+
+func TestPantheonBackupLinks(t *testing.T) {
```
@beeradb has stated he prefers a comment even for test functions :)

rfay commented on pkg/ddevapp/pantheon_test.go:
```diff
@@ -0,0 +1,130 @@
+
+	// testcommon.Chdir()() and CleanupDir() checks their own errors (and exit)
+	defer testcommon.Chdir(testDir)()
+	defer testcommon.CleanupDir(testDir)
```
Please swap order of defers.

rfay commented:
I did take this for a test drive on Windows, and I didn't have any trouble.

rfay commented:
If it's time for a merge on Monday and I'm not around to approve my changes, do feel free to just dismiss my review if you consider the changes made. Thanks!

rfay commented:
Per @beeradb I rebased this for testing purposes.

rfay commented:
@beeradb There's no sign that the TestPantheon* tests have been running during circle tests (or during mine on Windows). I don't know why this is. I think it's important to resolve before pull. See https://circleci.com/gh/drud/ddev/1348 for example.

beeradb commented:
With this last test run passing and @rfay giving the green light on windows, it's time to hit the button!
